WcaOnRails/app/views/registrations/_payment_form.html.erb

```diff
@@ -1,52 +1,133 @@
+    });
+    var elements = stripe.elements();
+    var cardElement = elements.create('card');
+    cardElement.mount('#card-element');
+    var cardholderName = document.getElementById('cardholder-name');
+    var $cardButton = $('#card-button');
 
-      $('#donate_button').on('click', function(e) {
-        e.preventDefault();
-        var registration_fees_to_pay = <%= @registration.outstanding_entry_fees.cents %>;
-        var donation = parseFloat($('#payment_donation').val() || 0);
-        var amount = parseFloat(registration_fees_to_pay+donation);
-        if (isNaN(amount)) {
-          alert('<%= t("registrations.payment_form.alerts.not_a_number") %>');
-        } else if (amount < registration_fees_to_pay) {
-          alert('<%= t("registrations.payment_form.alerts.amount_too_low") %>');
+    var setSaving = function() {
+      $cardButton.attr("disabled", true);
+      $cardButton.addClass("saving");
+    };
+    var unsetSaving = function() {
+      $cardButton.attr("disabled", false);
+      $cardButton.removeClass("saving");
+    };
+
+    function handleServerResponse(response) {
+      response.json().then(function (response_json) {
+        if (response_json.error) {
+          // Show error in payment form
+          $("#card-error").text(response_json.error.message);
+          unsetSaving();
+        } else if (response_json.requires_action) {
+          stripe.handleCardAction(
+            response_json.payment_intent_client_secret
+          ).then(function(result) {
+            if (result.error) {
+              $("#card-error").text(result.error.message);
+              unsetSaving();
+            } else {
+              // The card action has been handled
+              // The PaymentIntent can be confirmed again on the server
+              fetch('<%= registration_payment_intent_path(@registration) %>', {
+                method: 'POST',
+                headers: {
+                  'Content-Type': 'application/json',
+                  'X-CSRF-Token': document.querySelector('meta[name=csrf-token]').content
+                },
+                body: JSON.stringify({ payment_intent_id: result.paymentIntent.id })
+              }).then(handleServerResponse);
+            }
+          });
         } else {
-          $('#total_amount').val(amount);
-          handler.open({
-            amount: Math.round(amount)
-          })
+          // We got a successful payment, redirect to the registration page with a
+          // nice flash message.
+          window.location.replace('<%= competition_payment_success_path(@competition) %>');
         }
       });
-    })();
+    }
+
+    $cardButton.on('click', function(ev) {
+      ev.preventDefault();
+      setSaving();
+      var registration_fees_to_pay = <%= @registration.outstanding_entry_fees.cents %>;
+      var donation = parseFloat($('#payment_donation').val() || 0);
+      var amount = parseFloat(registration_fees_to_pay+donation);
```

Comment:
Ah right, I'm not sure what I was thinking initially, since this comes from the previous payment form.
I remember trying various amount-formatting library, maybe that was remaining from that.